File: frontend/types/configuracion.ts
/**
 * Configuración de reglas de negocio (migración 085).
 *
 * Todo resuelve por COALESCE(fila de mi empresa, fila global), y por eso cada bloque trae
 * `es_propia`: la pantalla necesita poder decir si estás mirando lo tuyo o lo heredado, porque
 * guardar mientras heredás CREA tu fila propia y te desengancha de la global.
 */

export interface Parametros {
  /** Denominador de la tasa de ausentismo. Era la constante 22 en el backend. */
  base_dias_habiles: number
  /** Mes contra el que se mide la antigüedad para ubicar al empleado en la escala. */
  corte_antiguedad_mes: number
  /**
   * Ventana en la que se pueden TOMAR vacaciones (p. ej. octubre→abril).
   *
   * ⚠️ CONCEPTO NUEVO Y HOY SOLO INFORMATIVO: no es cuándo se ganan los días ni cómo se mide
   * la antigüedad. El sistema GUARDA el valor y no lo aplica: NO hay validación que impida
   * cargar una licencia fuera de esta ventana, y no la habrá hasta definir si bloquea o solo
   * avisa. Ver el encabezado de la migración 085.
   */
  periodo_vacacional_desde_mes: number
  periodo_vacacional_hasta_mes: number
  /** En el primer año, mes desde el cual el ingreso otorga `primer_anio_dias`. */
  primer_anio_mes_corte: number
  primer_anio_dias: number
  /** Años que sobreviven los días no tomados antes de vencer. */
  vencimiento_anios: number
  /**
   * Migración 114. Días del período de prueba (LCT: 90). El "fin de período de prueba" de una
   * persona se calcula como `fecha_ingreso + periodo_prueba_dias`.
   *
   * ⚠️ HOY SOLO SE GUARDA Y SE MUESTRA: todavía no hay nada que lo calcule ni que avise. Está
   * expuesto porque es el valor que va a regir cuando eso se construya, y porque configurarlo
   * antes es gratis. Mismo estado que la ventana del período vacacional.
   */
  periodo_prueba_dias: number
  /**
   * Migración 114. Anticipación POR DEFECTO del aviso de un evento de agenda: con cuántos días
   * antes de la fecha aparece en el dashboard.
   *
   * 🔑 Es un DEFAULT, no una regla global: cada evento puede pisarlo con su propio `dias_aviso`.
   * Un evento que no lo trae toma este valor EN EL MOMENTO DEL ALTA y lo guarda como columna
   * propia, así que cambiarlo acá afecta a los eventos NUEVOS, no a los ya cargados.
   */
  dias_aviso_evento: number
}

export interface ParametrosResponse extends Parametros {
  es_propia: boolean
}

/** Un escalón: a partir de `antiguedad_anios` cumplidos corresponden `dias`. */
export interface TramoEscala {
  antiguedad_anios: number
  dias: number
}

export interface EscalaResponse {
  tramos: TramoEscala[]
  es_propia: boolean
}

export interface Configuracion {
  parametros: ParametrosResponse
  escala: EscalaResponse
}

File: backend/schemas/configuracion.py
"""
Schemas Pydantic de la configuración de reglas de negocio (migración 085).

Dos familias: los PARÁMETROS escalares y la ESCALA de vacaciones, que es una lista.
Las dos resuelven por COALESCE(fila de mi empresa, fila global), así que la salida lleva
`es_propia` para que la UI pueda decir si está mirando lo suyo o lo heredado.
"""
from typing import List, Optional

from pydantic import BaseModel, Field


class ParametrosUpdate(BaseModel):
    """
    Parámetros de la empresa activa. Todos obligatorios: es un PUT, no un PATCH — el form
    manda el juego completo y así no hay forma de guardar medio set y dejar el resto heredado
    sin que nadie lo note.

    Los rangos duplican los CHECK de la base a propósito: Pydantic devuelve un 422 con el
    campo señalado y la base un 500 genérico. El de la base es la red, no la validación.
    """

    base_dias_habiles: int = Field(ge=1, le=31)
    corte_antiguedad_mes: int = Field(ge=1, le=12)
    periodo_vacacional_desde_mes: int = Field(ge=1, le=12)
    periodo_vacacional_hasta_mes: int = Field(ge=1, le=12)
    primer_anio_mes_corte: int = Field(ge=1, le=12)
    primer_anio_dias: int = Field(ge=0)
    vencimiento_anios: int = Field(gt=0)
    # Migración 114. Los dos son de OTRA familia que los cinco de arriba —no son reglas de
    # vacaciones—, y viven en la misma tabla y en el mismo PUT a propósito: `parametros_empresa`
    # ya resuelve el patrón fila-global + fila-por-empresa-que-la-pisa, con sus dos UNIQUE
    # parciales. Una tabla aparte por familia habría duplicado ese mecanismo, y un upsert por
    # familia habría desenganchado a la empresa de las reglas globales de vacaciones al guardar
    # cualquiera de estos dos — que es exactamente el motivo por el que `parametros_screening`
    # sí es tabla propia (ahí el criterio lo escribe otro flujo, no este formulario).
    periodo_prueba_dias: int = Field(ge=1, le=730)
    # Anticipación por DEFECTO del aviso de un evento de agenda. Es el valor que toma un evento
    # nuevo que no trae `dias_aviso` propio; cada evento puede pisarlo. Ver schemas/evento_agenda.
    dias_aviso_evento: int = Field(ge=0, le=365)


class ParametrosResponse(ParametrosUpdate):
    """
    Los mismos valores más de dónde salieron.

    `es_propia=False` significa que la empresa NO tiene fila propia y está usando la global:
    la UI lo dice, porque editar y guardar en ese estado CREA la fila propia y a partir de ahí
    la empresa deja de seguir a la global. Sin ese aviso, el primer guardado desengancharía la
    empresa sin que nadie lo haya decidido.
    """

    es_propia: bool


class TramoEscala(BaseModel):
    """Un escalón: a partir de `antiguedad_anios` cumplidos corresponden `dias`."""

    antiguedad_anios: int = Field(ge=0, le=60)
    dias: int = Field(gt=0, le=365)


class EscalaResponse(BaseModel):
    tramos: List[TramoEscala]
    es_propia: bool


class EscalaUpdate(BaseModel):
    """
    La escala COMPLETA de la empresa activa. Se reemplaza entera, no se parchea tramo a tramo:
    la operación de negocio es "esta es mi escala", y un alta/baja por tramo dejaría estados
    intermedios visibles (una escala sin el tramo de 0 años no aplica a nadie nuevo).
    """

    tramos: List[TramoEscala]


class ConfiguracionResponse(BaseModel):
    """Todo lo que la pantalla de configuración necesita, en una sola ida."""

    parametros: ParametrosResponse
    escala: EscalaResponse


class TipoAusenciaUpdate(BaseModel):
    """
    Edición de un tipo de ausencia. Todo opcional: es un PATCH real, se manda solo lo que
    cambia. `activo=False` es la BAJA — no hay DELETE, ver tipos_ausencia_service.
    """

    nombre: Optional[str] = None
    activo: Optional[bool] = None
    cuenta_ausentismo: Optional[bool] = None
